tests/evm_backends/pyevm_env.py

```diff
@@ -0,0 +1,222 @@
+from eth.chains.mainnet import MainnetChain
+from eth.constants import CREATE_CONTRACT_ADDRESS, GENESIS_DIFFICULTY
+from eth.db.atomic import AtomicDB
+from eth.exceptions import Revert, VMError
+from eth.tools.builder import chain as chain_builder
+from eth.vm.base import StateAPI
+from eth.vm.execution_context import ExecutionContext
+from eth.vm.message import Message
+from eth.vm.transaction_context import BaseTransactionContext
+from eth_keys.datatypes import PrivateKey
+from eth_typing import Address
+from eth_utils import setup_DEBUG2_logging, to_canonical_address, to_checksum_address
+
+import vyper.evm.opcodes as evm_opcodes
+from tests.evm_backends.base_env import BaseEnv, EvmError, ExecutionResult, LogEntry
+from vyper.utils import keccak256
+
+
+class PyEvmEnv(BaseEnv):
+    """EVM backend environment using the Py-EVM library."""
+
+    INVALID_OPCODE_ERROR = "Invalid opcode"
+
+    def __init__(
+        self,
+        gas_limit: int,
+        account_keys: list[PrivateKey],
+        tracing=False,
+        block_number=1,
+        evm_version="mainnet",
+    ) -> None:
+        super().__init__(gas_limit, account_keys)
+
+        # note: we configure the evm version that we emit code for,
+        # but eth-tester is only configured with the latest mainnet
+        # version.
+        evm_opcodes.DEFAULT_EVM_VERSION = evm_version
+
+        if tracing:
+            logger = logging.getLogger("eth.vm.computation.BaseComputation")
+            setup_DEBUG2_logging()
+            logger.setLevel("DEBUG2")
+            # from vdb import vdb
+            # vdb.set_evm_opcode_debugger()
+
+        spec = getattr(chain_builder, evm_version + "_at")(block_number)
+        self._chain: ChainAPI = chain_builder.build(MainnetChain, spec).from_genesis(
+            base_db=AtomicDB(),
+            genesis_params={"difficulty": GENESIS_DIFFICULTY, "gas_limit": gas_limit},
+        )
+
+        self._last_computation: ComputationAPI = None
+
+    @cached_property
+    def _state(self) -> StateAPI:
+        return self._vm.state
+
+    @cached_property
+    def _vm(self):
+        return self._chain.get_vm()
+
+    @cached_property
+    def _context(self) -> ExecutionContext:
+        context = self._state.execution_context
+        assert isinstance(context, ExecutionContext)
+        return context
+
+    @contextmanager
+    def anchor(self):
+        snapshot_id = self._state.snapshot()
+        try:
+            yield
+        finally:
+            self._state.revert(snapshot_id)
+
+    def get_balance(self, address: str) -> int:
+        return self._state.get_balance(_addr(address))
+
+    def set_balance(self, address: str, value: int):
+        self._state.set_balance(_addr(address), value)
+
+    @property
+    def accounts(self) -> list[str]:
+        return [key.public_key.to_checksum_address() for key in self._keys]
+
+    @property
+    def block_number(self) -> int:
+        return self._state.block_number
```

Comment:
iirc this just returns context.block_number, for consistency (e.g. with time_travel) we should use `self._context` though